### ISSUE TITLE
 IWD2: Chult portal timer elapsed before killing the guardian

### ISSUE THREAD

**Issue by MarcelHB** [gemrb/gemrb#2184]
#### Bug description
I was suddenly greeted by the message that the game has ended for me while the dragon was only half-way down.

From how it's supposed to go: guardian dies -> countdown to reach portal. And if you fail, then you are stuck. But apparently the countdown starts at some earler point before death, or something messed up during the fight.

#### Screenshots
![iwd2_chult_end](https://github.com/user-attachments/assets/5306373c-b8e4-462e-a557-827596b44bbd)

GemRB version (check as many as you know apply)
- [x] master as of this issue
- [ ] 0.9.3
- [ ] 0.9.2
- [ ] 0.9.1

**Comment by lynxlynxlynx** [comment]
What's messed up is that you changed areas.  It shouldn't matter though, since TimerPortalClose and the vars are all global. 60splfx.baf is the only thing starting the chain (Chult_PortalClosing -> 1), only once the dragon is dead.

I never had such problems with the dragon though and it doesn't really make sense. Can you reproduce at all?

**Comment by MarcelHB** [comment]
I have a savegame after beginning the fight for which this is reproducible.

By "messed up" what's the original problem then?

**Comment by lynxlynxlynx** [comment]
Stand your ground and fight. :) I don't see any code for disabling the entrance, so it was probably possible in the original as well.

In the save Chult_PortalClosing is already 1, right? It's useless then.

**Comment by MarcelHB** [comment]
Ok, I'll try to find the time with some earlier savegame and try again.

**Comment by MarcelHB** [comment]
Can reproduce this with an earlier savegame.

Looks as if `Dead` check fails when the dragon has left AR6500 into AR6501, around here: https://github.com/gemrb/gemrb/blob/c735112d670db744e06dc33c85fc5e36ee9ccaf1/gemrb/core/GameScript/Triggers.cpp#L1800-L1804

Maybe this thing should also consider the whole map cluster instead of just the one that hosts the script?

**Comment by lynxlynxlynx** [comment]
Are you sure you're looking at the correct call? `Dead("The_Guardian")` shouldn't reach that part of the function.

**Comment by MarcelHB** [comment]
Yes, `actor` was null here but I got that `parameter->objectName` to reveal `"The_Guardian"`

**Comment by lynxlynxlynx** [comment]
Ah, yes, IESDP has a bug, it always uses an object parameter (... fixed). I think we can safely turn that 1 into 0, however upon checking the original I see it only returns 0 if the object is found, is an actor and has the dead state set in neither Modified nor BaseStats. So I think we should still do it, you just found a bug in the original. :)

**Comment by MarcelHB** [comment]
Are you looking at https://github.com/alexbatalov/iwd2-re or is there anything more comprehensive?

**Comment by lynxlynxlynx** [comment]
Disassembly — I don't think iwd2-re has this covered yet.

**Comment by MarcelHB** [comment]
Are there any PDB files or annotations around? Could check something like this myself, but the only clues I find in my IWD2.exe are hours away from any meaningful conclusion. 😄

**Comment by lynxlynxlynx** [comment]
I got Bubb's ghidra files to map out the Door stuff and then did some other random work while exploring. IIRC the PPC port was the only one with debug symbols and I couldn't find that anywhere any more, and the build was optimized, so it doesn't decompile very cleanly.

**Comment by lynxlynxlynx** [comment]
The original doesn't close the entrance, nor does it follow. I can't reproduce this in gemrb either and I tried with a gimped script where he's ordered to just attack or MoveToObject. Either that trigger fix was enough (for a clean game), other improvements fixed it or you were tremendously unlucky and somehow the teleport resolved into the new area. Closing.
